### PR TITLE
Ability to change uri for query

### PR DESCRIPTION
**Pull request by 4n70w4**

### CODE REVIEW

**Comment by rccrdpccl** [comment]
Thanks for the contribution @4n70w4 .

Could you please explain your use case behind this, so that we could better understand?

Could you also please add a simple test case?

File: src/Client.php [hunk 1]
```diff
@@ -17,10 +17,14 @@ public function __construct(ClientInterface $httpClient, ResponseBuilder $respon
     }
 
     /**
+     * @param string $query
```
Comment (rccrdpccl) [inline src/Client.php:20]: Code standards in this project are to not declare variables in annotations: we use type hinting instead, and use annotations only for exceptions that unfortunately cannot be declared in the function's signature. Could you please remove annotations for params and return value?

File: src/Client.php [hunk 2]
```diff
@@ -17,10 +17,14 @@ public function __construct(ClientInterface $httpClient, ResponseBuilder $respon
...
+     * @param array|null $variables
+     * @param $uri
+     * @return Response
      * @throws \UnexpectedValueException When response body is not a valid json
      * @throws \RuntimeException         When there are transfer errors
      */
-    public function query(string $query, array $variables = null): Response
+    public function query(string $query, array $variables = null, $uri = ''): Response
```
Comment (rccrdpccl) [inline src/Client.php:27]: Could you please add type hinting to the newly added parameter?

**Comment by 4n70w4** [comment]
@rccrdpccl example:

I reuse the already configured guzzle client and needs to define the endpoint for the request.

About the tests, this does not affect your library. Also does not create new mutants in mutation testing.

**Comment by rccrdpccl** [comment]
Isn't this configurable by passing guzzle options to the client? I am curious to understand in what case you'd need to redefine request uri from already initialized client.

Tests are not only for the code that is already there, but it prevents future changes to break the functionality that you are introducing. However we can wait to define exact use case before writing tests.

**Comment by 4n70w4** [comment]
```php

use Softonic\GraphQL\Client;
use Softonic\GraphQL\Response;
use Softonic\GraphQL\ResponseBuilder;

....

    public function something(array variables, $query, $uri) : Response {
        $client = new Client($this->guzzle, new ResponseBuilder() );

        return $client->query($query, $variables, $uri);
    }

....

// another methods to use guzzle client for non graphql requests

```

**Comment by rccrdpccl** [comment]
Thanks for the example. What is the reason for re-using guzzle client?

**Comment by joskfg** [comment]
Thanks @4n70w4 for your PR!

I think add the uri in the query call could be not the best option. We could find a better solution if we know the exact use case to do something cleaner if it is possible.

If the case is to access two different graphql servers I think that the best option is to create two clients. This make sense because probably the clients should have different options. For example, the authentication could be different.

**Comment by 4n70w4** [comment]
In your opinion hardcode uri as a blank line is the best option?

**Comment by 4n70w4** [comment]
@rccrdpccl I already have a configured client (authorization, certificates, proxies, etc.) for access to the provider’s endpoints. I see no reason to configure another client with a difference only in the url '/ graphql'.

**Comment by rccrdpccl** [comment]
@4n70w4 Ok, thanks for explaining, it seems a reasonable use case. I think your approach it's a good improvement for the library, ideally we should end up with something like `public function query(string $query, array $variables = null, string $uri = '', array $options = []): Response`, but for now we can add parameter `$uri` and leave options to be added in the near future.

I will add a few small comments to the PR so we can get it merged as soon as possible!

In order to solve this long term, we are planning to release `v2`, where we will need to configure the client in a similar fashion: `new Client($this->guzzle, new ResponseBuilder(), '/quert_endpoint', '/mutation_endpoint' );`. Would this work for your use case?

**Review by rccrdpccl — CHANGES_REQUESTED** [review]
Please do not forget to add tests, so that we make sure we will not have any regression on this behaviour in the future.

**Comment by 4n70w4** [comment]
I already use my fork in production. I don’t know if it will be time to go back to this PR.

Such a design could also solve similar problems:

```php
new Client($this->guzzle, new ResponseBuilder(), '/endpoint');
```

**Comment by rccrdpccl** [comment]
Close for inactivity. If interested in implementing this issue, please re-open it or file an issue

**Pull request closed without merging**